x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -533,22 +542,27 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
-                final boolean reloadBefore = this.contentAndFilteredJwksAlgsHmac.isEmpty() && this.contentAndFilteredJwksAlgsPkc.isEmpty();
-                final ContentAndFilteredJwksAlgs contentAndFilteredJwksAlgs = isJwtSignatureAlgHmac
-                    ? this.contentAndFilteredJwksAlgsHmac
-                    : this.contentAndFilteredJwksAlgsPkc;
                 try {
-                    if ((reloadBefore) && (this.reloadJwksHelper(listener, tokenPrincipal, isJwtSignatureAlgHmac) == false)) {
-                        return; // Reload required before. Reload failed, or succeeded but no changes. Helper populates listener response.
+                    if (this.contentAndFilteredJwksAlgsHmac.isEmpty() && this.contentAndFilteredJwksAlgsPkc.isEmpty()) {
+                        LOGGER.debug("Realm [" + super.name() + "] has no JWKs and Algs to verify JWT for token=[" + tokenPrincipal + "].");
+                        if (this.reloadJwksHelper(listener, tokenPrincipal, isJwtSignatureAlgHmac) == false) {
+                            return; // Try reload before. Reload failed, or succeeded but no changes. Helper handles listener and logs.
+                        }
                     }
-                    // Throws AllVerifiesFailedException (some JWKs after filtering, all failed) or Exception (no JWKs after filtering)
+                    // Throws Exception (no JWKs after filtering) or AllVerifiesFailedException (all available JWKs failed)
+                    final ContentAndFilteredJwksAlgs contentAndFilteredJwksAlgs = isJwtSignatureAlgHmac
+                        ? this.contentAndFilteredJwksAlgsHmac
+                        : this.contentAndFilteredJwksAlgsPkc;
                     JwtValidateUtil.validateSignature(jwt, contentAndFilteredJwksAlgs.filteredJwksAlgs.jwks);
                     // Fall through means success, exception will be caught by inner or outer catch
                 } catch (JwtValidateUtil.AllVerifiesFailedException e) {
-                    if ((reloadBefore == false) && (this.reloadJwksHelper(listener, tokenPrincipal, isJwtSignatureAlgHmac) == false)) {
-                        return; // Reload required after. Reload failed, or succeeded but no changes. Helper populates listener response.
+                    if (this.reloadJwksHelper(listener, tokenPrincipal, isJwtSignatureAlgHmac) == false) {
+                        return; // Try reload before. Reload failed, or succeeded but no changes. Helper handles listener and logs.
                     }
-                    // Throws AllVerifiesFailedException (some JWKs after filtering, all failed) or Exception (no JWKs after filtering)
+                    // Throws Exception (no JWKs after filtering) or AllVerifiesFailedException (all available JWKs failed)
+                    final ContentAndFilteredJwksAlgs contentAndFilteredJwksAlgs = isJwtSignatureAlgHmac
```

Comment:
These members are replaced reloadJwksHelper, so get the new reference for retrying signature validation with the new JWKs.